inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,35 +198,70 @@ def post_audiences(self, course, data, active_tab, msg, error):
+                auth_method_ids = self.user_manager.get_auth_methods().keys()
+                students_per_audience = {}
+                tutors_per_audience = {}
+                course_students = []
+                course_tutors = []
+                audiences = []
+                for line in csv_data:
+                    if len(line) != 4:
+                        msg["audiences"] = _("File wrongly formatted.")
+                        error["audiences"] = True
+                if "audiences" not in error or not error["audiences"]:
+                    stud_list, aud_li, oth_stu, u_info = self.get_user_lists(course)
+                    courseid = course.get_id()
+                    self.database.audiences.delete_many({"courseid": courseid})
+                    for user_id, field, role, description in csv_data:
+                        user_id = user_id.strip()
+                        field = field.strip()
+                        role = role.strip()
+                        description = description.strip()
+                        if field not in ["username", "email", "realname"] + list(auth_method_ids) \
+                                or role not in ["student", "tutor"]:
+                            msg["audiences"] = _("Some fields were not recognized:" + field)
+                            error["audiences"] = True
+                            continue
+                        if field != "username":
+                            user = self.database.users.find_one({field: user_id})
+                            if user is not None:
+                                user_id = user["username"]
+                            else:
+                                msg["audiences"] = _("Some users were not found: " + user_id)
+                                error["audiences"] = True
+                                continue
+                        if role == "student":
+                            students_per_audience.setdefault(description, []).append(user_id)
+                            course_students.append(user_id)
+                        else:
+                            tutors_per_audience.setdefault(description, []).append(user_id)
+                            course_tutors.append(user_id)
+                    for key, value in students_per_audience.items():
+                        audiences.append({"description": key, "courseid": courseid,
+                                  "students": value, "tutors": tutors_per_audience[key] if key in tutors_per_audience else []})
+                    new_students = list(set(stud_list).union(set(course_students)))
+                    new_tutors = list(set(course.get_tutors()).union(set(course_tutors)))
+                    self.database.courses.update_one({"_id": "LFSAB1106"}, {"$set": {"students": new_students,
+                                                                                     "tutors": new_tutors}})
+                    for audience in audiences:
```

Comment:
There should be a comment here to specify that this is done to avoid removing the audience id and impact the group audience filter. If anyone forgets about this, they may remove the code.